[fix] Improve ownership handling and documentation for Stacked Diagrams

gacarrillor opened the pull request:
Followup #58568 (probably the last one in this series)

Improves ownership handling of sub renderers in `QgsStackedDiagramRenderer` and `QgsStackedDiagramPropertiesModel`.

Adds documentation on ownership transfer when receiving or returning sub renderers.

Other adjustments:
 + Replaces `subDiagram` by `subDiagramRenderer` when we actually refer to renderers.
 + Avoids `auto` for non-complex types.
 + Passing from stacked to single diagram: take first stacked diagram as the basis for the new single one.
 + Stacked diagram list in GUI: Match `Diagram type` column to names from the main diagram combobox. This helps us to better differentiate between a stacked bars diagram and a stacked diagram (using the diagram name alone was a bit confusing: e.g., `Stacked`).

gacarrillor committed:
[core] Make sure QgsStackedDiagramRenderer::addRenderer() takes ownership

gacarrillor committed:
[core] Implement StackedDiagramRenderer's copy constructor (called by clone()) and operator=

gacarrillor committed:
[gui] Document and handle ownership in QgsStackedDiagramProperties and underlying model

gacarrillor committed:
[core] Remove unused method in QgsStackedDiagramRenderer

nyalldawson commented on src/gui/vector/qgsstackeddiagramproperties.cpp:
```diff
@@ -189,7 +189,7 @@ void QgsStackedDiagramProperties::syncToLayer()
       const auto renderers = stackedDiagramRenderer->renderers();
       for ( const auto &renderer : renderers )
```
Can you avoid the use of auto for things like this? It goes against current QGIS coding guidelines, which is only to use auto for complex types like iterators.

gacarrillor replied:
Sure! Thanks for the suggestion.

gacarrillor replied:
Addressed in https://github.com/qgis/QGIS/pull/59123/commits/3bbad5495b61cbc8bc297c01f6c8cce831ef8bd8

nyalldawson reviewed (approved):
Looks good, thanks!

gacarrillor commented:
@nyalldawson, do you know how to get rid of the error in code layout?

```
ERROR: Issues in 'warning' category found:
/home/runner/work/QGIS/QGIS/src/core/qgsdiagramrenderer.cpp:867,warning,missingMemberCopy,Member variable 'QgsStackedDiagramRenderer::mDiagramRenderers' is not assigned in the copy constructor. Should it be copied?
```

gacarrillor committed:
[gui] Rename subDiagram related methods to subDiagramRenderer in QgsStackedDiagramProperties

nyalldawson commented:
@gacarrillor 
> do you know how to get rid of the error in code layout?

Add an explicit empty constructor for  mDiagramRenderers -- that indicates that you're intentionally NOT copying that field

gacarrillor committed:
Address review: Avoid autos in stacked diagrams code

gacarrillor committed:
[fix] Avoid code layout error by adding an explicit empty constructor for mDiagramRenderers, indicating that we won't copy that field

gacarrillor committed:
[fix] When passing from Stacked Diagram to a single one, take the first one (if it exists) in the stacked diagram as a basis for the new single one

Before this commit, the stacked diagram properties were taken as a basis, losing valuable diagram settings in the process.
Note this reverts commit 2f9a0758b309ef162d6390b7411cba71af3375be

gacarrillor committed:
[ux] Stacked diagram list in GUI: Match Diagram type column to names from the main diagram combobox

This helps us to better differentiate between a stacked bars diagram and a stacked diagram (using the diagram name alone was a bit confusing: e.g., Stacked).

gacarrillor commented:
@nyalldawson,
Sorry, I added two last commits to sort of close this followup series.
Thanks for your review!
